### ISSUE TITLE
Expose support for asynchronous interactions

### ISSUE THREAD

**Issue by jboynes** [GoogleCloudPlatform/google-cloud-java#63]
Servlet 3.1 and frameworks like Spring 4 support non-blocking request execution. This partially supported by the raw Storage API which exposes its HttpRequest objects with their executeAsync() operation. However, this does not support non-blocking transfers.

Full async support would allow an application to transfer data to/from Cloud Storage without having to consume a Thread while waiting for IO to complete.

**Comment by aozarov** [comment]
I doubt that will change while using google-api-java-client (did you create a feature request there we can link too?), so the current plan is to support it for every service that is available via gRPC.

**Comment by aozarov** [comment]
This should not be a [Needed for public announcement](https://github.com/GoogleCloudPlatform/gcloud-java/milestones/Needed%20for%20public%20announcement) blocker but rather more of a future enhancement (when service supports it).

**Comment by jgeewax** [comment]
I can buy that. We could add these later pretty easily. Moving out the of the milestone.

**Comment by susanlinsfu** [comment]
+1

**Comment by lesv** [comment]
Turns out that AppEngine Standard has [async support](https://cloud.google.com/appengine/docs/java/issue-requests#issuing_an_asynchronous_request), Jetty 3.1 in Flex supports Async (at Top of Trunk), for generic Java, there is [Apache's HttpAsyncClient](https://hc.apache.org/httpcomponents-asyncclient-dev/quickstart.html)

**Comment by garrettjonesgoogle** [comment]
I think we should just naturally surface the async functionality as we move each API to gRPC instead of having an old tracking issue. Closing this out.

**Comment by dminkovsky** [comment]
Has this been addressed elsewhere? I'm not finding another issue with this topic.

What is currently the best way to stream a file from storage with Netty? As in https://github.com/netty/netty/blob/60250f3795d4082e59df6806dade2966ec583859/example/src/main/java/io/netty/example/http/file/HttpStaticFileServerHandler.java#L188-L202.

**Comment by dminkovsky** [comment]
The way I have this working now is:

```
            ReadChannel reader = storage.reader("bucket", "blob");
            InputStream in = Channels.newInputStream(reader);
            ChunkedStream stream = new ChunkedStream(in);
            HttpChunkedInput input = new HttpChunkedInput(stream);
```

I send the `HttpChunkedInput` stream through the Netty `ChuckedWriteHandler` writes it out.

I wonder, though: the ReadChannel reader, is it non-blocking?

**Comment by oripwk** [comment]
Is there anything new with this?

Should we still work with GCS in a fully blocking manner? Or is there a solution?

Thanks

**Comment by frankyn** [comment]
@oripwk we don't currently have a plan for async in the GCS library. Could you file an issue against https://github.com/googleapis/java-storage/issues so we don't lose track of your request? Thank you!

**Comment by oripwk** [comment]
@frankyn opened here: https://github.com/googleapis/java-storage/issues/638

**Comment by agentgt** [comment]
For download you only need to block on generating the download url. e.g: `Storage.signUrl`

Then you can use whatever Http client that is non blocking. 

`Storage.signUrl` on code inspection appears to be non blocking as there doesn't seem to be any IO.

**Comment by danielcompton** [comment]
This is available now in [v2.2.3](https://github.com/tfutils/tfenv/releases/tag/v2.2.3).
